libraries/extractor/src/main/java/androidx/media3/extractor/ts/DtsUhdReader.java

```diff
@@ -0,0 +1,246 @@
+  @Nullable private final String language;
+
+  private @MonotonicNonNull String formatId;
+  private @MonotonicNonNull TrackOutput output;
+
+  private int state;
+  private int bytesRead;
+
+  // Used to find the header.
+  private int syncBytes;
+
+  // Used when parsing the header.
+  private long sampleDurationUs;
+  private @MonotonicNonNull Format format;
+  private int sampleSize;
+  private boolean isFtocSync;
+  private int headerSizeToRead;
+
+  // Used when reading the samples.
+  private long timeUs;
+  private int sampleRate;
+  private int sampleCount; // frame duration
+
+  /**
+   * Constructs a new reader for DTS UHD elementary streams.
+   *
+   * @param language Track language.
+   */
+  public DtsUhdReader(@Nullable String language) {
+    headerScratchBytes = new ParsableByteArray(new byte[FTOC_HEADER_SIZE_MAX]);
+    state = STATE_FINDING_SYNC;
+    timeUs = C.TIME_UNSET;
+    sampleRate = 48000; // initialize to a non-zero sampling rate
+    this.language = language;
+  }
+
+  @Override
+  public void seek() {
+    state = STATE_FINDING_SYNC;
+    bytesRead = 0;
+    syncBytes = 0;
+    timeUs = C.TIME_UNSET;
+  }
+
+  @Override
+  public void createTracks(ExtractorOutput extractorOutput, TrackIdGenerator idGenerator) {
+    idGenerator.generateNewId();
+    formatId = idGenerator.getFormatId();
+    output = extractorOutput.track(idGenerator.getTrackId(), C.TRACK_TYPE_AUDIO);
+  }
+
+  @Override
+  public void packetStarted(long pesTimeUs, @TsPayloadReader.Flags int flags) {
+    if (pesTimeUs != C.TIME_UNSET) {
+      timeUs = pesTimeUs;
+    }
+  }
+
+  @Override
+  public void consume(ParsableByteArray data) {
+    Assertions.checkStateNotNull(output); // Asserts that createTracks has been called.
+    while (data.bytesLeft() > 0) {
+      switch (state) {
+        case STATE_FINDING_SYNC:
+          if (skipToNextSync(data)) {
+            state = STATE_FIND_HEADER_SIZE;
+          }
+          break;
+        case STATE_FIND_HEADER_SIZE:
+          // Read enough bytes to parse the header size information.
+          if (continueRead(data, headerScratchBytes.getData(), FTOC_HEADER_SIZE_MIN)) {
+            findHeaderSize();
+            state = STATE_READING_HEADER;
+          }
+          break;
+        case STATE_READING_HEADER:
+          if (continueRead(data, headerScratchBytes.getData(), headerSizeToRead)) {
+            parseHeader();
+            headerScratchBytes.setPosition(0);
+            output.sampleData(headerScratchBytes, headerSizeToRead);
+            state = STATE_READING_SAMPLE;
+          }
+          break;
+        case STATE_READING_SAMPLE:
+          int bytesToRead = min(data.bytesLeft(), sampleSize - bytesRead);
+          output.sampleData(data, bytesToRead);
+          bytesRead += bytesToRead;
+          if (bytesRead == sampleSize) {
+            if (timeUs != C.TIME_UNSET) {
+              output.sampleMetadata(timeUs, C.BUFFER_FLAG_KEY_FRAME, sampleSize, 0, null);
```

Comment:
Is this correct? what if it's a frame with syncWord == SYNC_VALUE_UHD_FTOC_NONSYNC_BE? Based on 103 491 section 6.4.4.2 we should only mark frames with syncWord == SYNC_VALUE_UHD_FTOC_SYNC_BE as keyframes.

Reply:
The FTOC sync and non-sync frames need to pass to the DTS decoder. 
Which flag do we need to use with the non-sync frame if the BUFFER_FLAG_KEY_FRAME is not suitable for such frames?

Reply:
Agreed that in normal cases we need to feed all (sync and non-sync) frames to the decoder. 

I don't think we can start decoding from a non-sync frame and as such non-sync frames should be marked as `BUFFER_FLAG_KEY_FRAME`. Please also check if changing this would result in test failures while seeking.

Reply:
The decoding won't start from the non-sync frame. In the seek() function, we are setting 'uhdInSync' to false.
The 'skipToNextSync()' function will return 'true' only if "it is a FTOC sync frame" OR "FTOC non-sync frame and already parsed a FTOC sync".